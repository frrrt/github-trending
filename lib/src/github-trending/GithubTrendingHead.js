import React from 'react';
import Head from 'next/head';

const GithubTrendingHead = () => (
    <Head>
        <title>Github Trending</title>
        <meta name="description" content="Github Trending shows you the most popular repos of the last week and lets you star them!" />
    </Head>
);

export default GithubTrendingHead;
