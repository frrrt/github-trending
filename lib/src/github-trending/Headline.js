import React from 'react';
import PropTypes from 'prop-types';

import { Typography, Box } from '@material-ui/core';

const Headline = ({ children }) => (
    <Box mt={4}>
        <Typography variant="h2" component="h1">{children}</Typography>
    </Box>
);

Headline.propTypes = {
    children: PropTypes.node.isRequired,
};

export default Headline;
