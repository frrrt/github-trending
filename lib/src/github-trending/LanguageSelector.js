import React from 'react';
import PropTypes from 'prop-types';

import Autocomplete from '@material-ui/lab/Autocomplete';
import { TextField, makeStyles } from '@material-ui/core';

const filterAvailableLanguages = currentLanguages => currentLanguages
    .sort()
    // Filter duplicate entries, from stackoverflow
    .filter((item, pos, ary) => !pos || item !== ary[pos - 1]);

const useStyles = makeStyles(() => ({
    root: {
        // width: 200 ensures that layout works on iPhone5 sized screens
        width: 200,
    },
}));

const LanguageSelector = ({ onChange, currentLanguages }) => {
    const classes = useStyles();

    return (
        <Autocomplete
            id="language-selector"
            options={filterAvailableLanguages(currentLanguages)}
            getOptionLabel={option => option}
            className={classes.root}
            size="small"
            selectOnFocus
            clearOnBlur
            handleHomeEndKeys
            // eslint-disable-next-line react/jsx-props-no-spreading
            renderInput={params => <TextField {...params} label="Language" variant="outlined" />}
            onChange={(_, newValue) => onChange(newValue)}
        />
    );
};

LanguageSelector.propTypes = {
    onChange: PropTypes.func.isRequired,
    currentLanguages: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default LanguageSelector;
