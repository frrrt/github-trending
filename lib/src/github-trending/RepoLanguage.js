import React from 'react';
import PropTypes from 'prop-types';

import { makeStyles } from '@material-ui/core';
import LanguageColors from '../../languageColors';

const useStyles = makeStyles(theme => ({
    root: {
        display: 'flex',
        alignItems: 'center',
    },
    circle: {
        borderRadius: theme.spacing(1),
        width: theme.spacing(2),
        height: theme.spacing(2),
        marginRight: theme.spacing(1),
    },
}));

const RepoLanguage = ({ language }) => {
    const classes = useStyles();

    return (
        <div className={classes.root}>
            <div className={classes.circle} style={{ backgroundColor: LanguageColors[language] ?? (language && '#ccc') }} />
            {language}
        </div>
    );
};

RepoLanguage.propTypes = {
    language: PropTypes.string,
};

RepoLanguage.defaultProps = {
    language: '',
};

export default RepoLanguage;
