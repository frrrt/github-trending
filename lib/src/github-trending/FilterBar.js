import React from 'react';
import PropTypes from 'prop-types';

import {
    Box,
    Button,
    makeStyles,
} from '@material-ui/core';

import FilterIcon from '@material-ui/icons/FilterList';

import LanguageSelector from './LanguageSelector';

const useStyles = makeStyles(theme => ({
    root: {
        display: 'flex',
        alignItems: 'center',
    },
    starredButton: {
        marginLeft: '2rem',
        whiteSpace: 'no-wrap',
    },
    starredIcon: {
        marginRight: theme.spacing(1),
    },
}));

const FilterBar = ({
    isStarredFiltered,
    availableLanguages,
    onChangeLanguage,
    onChangeStarred,
}) => {
    const classes = useStyles();

    return (
        <Box mt={4} className={classes.root}>
            <LanguageSelector
                currentLanguages={availableLanguages}
                onChange={onChangeLanguage}
            />

            <Button
                variant="outlined"
                className={classes.starredButton}
                onClick={onChangeStarred}
                color={isStarredFiltered ? 'secondary' : undefined}
            >
                <FilterIcon className={classes.starredIcon} /> Starred
            </Button>
        </Box>
    );
};

FilterBar.propTypes = {
    isStarredFiltered: PropTypes.bool.isRequired,
    availableLanguages: PropTypes.arrayOf(PropTypes.string).isRequired,
    onChangeLanguage: PropTypes.func.isRequired,
    onChangeStarred: PropTypes.func.isRequired,
};

export default FilterBar;
