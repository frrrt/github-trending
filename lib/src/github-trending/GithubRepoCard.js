import React from 'react';
import PropTypes from 'prop-types';

import {
    Typography,
    Link,
    Card,
    CardContent,
    makeStyles,
} from '@material-ui/core';

import RepoLanguage from './RepoLanguage';
import StarButton from './StarButton';
import githubRepoPropType from '../../proptypes/githubRepoPropType';

const useStyles = makeStyles(() => ({
    root: {
        flexGrow: 1,
    },
    repoTitle: {
        // prevent very long repo names from braking layout
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        textOverflow: 'ellipsis',
    },
    main: {
        display: 'flex',
        justifyContent: 'space-between',
        // TODO: height 31 mitigates content jumping which comes from a bug in mui
        height: 31,
    },
}));

const GithubRepoCard = ({ repo, isStarred, onToggleStar }) => {
    const classes = useStyles();

    return (
        <Card className={classes.root}>
            <CardContent>
                <Typography gutterBottom variant="h6" component="div" className={classes.repoTitle}>
                    {/* title attribute is important for very long repo names which get cut */}
                    {/* noopener noreferrer may be obsolete: see https://web.dev/external-anchors-use-rel-noopener/ */ }
                    <Link href={repo.html_url} title={repo.name} target="_blank" rel="noopener noreferrer">
                        {repo.name}
                    </Link>
                </Typography>
                <Typography gutterBottom variant="body2" component="div" className={classes.main}>
                    <RepoLanguage language={repo.language} />

                    <StarButton
                        starsCount={repo.watchers}
                        isStarred={isStarred}
                        onToggleStar={onToggleStar}
                    />
                </Typography>
                <Typography gutterBottom variant="body2">
                    {repo.description}
                </Typography>
            </CardContent>
        </Card>
    );
};

GithubRepoCard.propTypes = {
    repo: githubRepoPropType.isRequired,
    isStarred: PropTypes.bool.isRequired,
    onToggleStar: PropTypes.func.isRequired,
};

export default GithubRepoCard;
