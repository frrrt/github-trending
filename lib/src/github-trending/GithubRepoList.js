import React from 'react';
import PropTypes from 'prop-types';
import {
    Typography,
    Box,
    Grid,
    makeStyles,
} from '@material-ui/core';

import githubRepoPropType from '../../proptypes/githubRepoPropType';
import GithubRepoCard from './GithubRepoCard';

const useStyles = makeStyles(() => ({
    row: {
        // flex box ensures, that the cards in a row have the same height
        // disregarding the natural height from their innenr content.
        display: 'flex',
    },
}));

const GithubRepoList = ({ filteredRepos, starred, onToggleStar }) => {
    const classes = useStyles();

    return (
        <Box mt={4} mb={8}>
            <Grid container spacing={2}>
                {filteredRepos.map(repo => (
                    <Grid key={repo.id} item xs={12} sm={6} md={4} lg={4} className={classes.row}>
                        <GithubRepoCard
                            repo={repo}
                            onToggleStar={() => onToggleStar(repo.id)}
                            isStarred={Boolean(starred[repo.id])}
                        />
                    </Grid>
                ))}
            </Grid>
            {filteredRepos.length === 0 && (
                <Typography variant="body1">
                    No repos found!
                </Typography>
            )}
        </Box>
    );
};

GithubRepoList.propTypes = {
    filteredRepos: PropTypes.arrayOf(githubRepoPropType).isRequired,
    onToggleStar: PropTypes.func.isRequired,
    starred: PropTypes.shape({}).isRequired,
};

export default GithubRepoList;
