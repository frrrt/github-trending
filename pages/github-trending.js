import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { Container } from '@material-ui/core';

import { format, addWeeks } from 'date-fns';

import { toggleStar } from '../store/actions';
import githubRepoPropType from '../lib/proptypes/githubRepoPropType';
import Headline from '../lib/src/github-trending/Headline';
import FilterBar from '../lib/src/github-trending/FilterBar';
import GithubRepoList from '../lib/src/github-trending/GithubRepoList';
import GithubTrendingHead from '../lib/src/github-trending/GithubTrendingHead';

export const GithubTrendingPage = ({
    repos,
    starred,
    onToggleStar,
}) => {
    const [language, setLanguage] = useState(null);
    const [isStarredFiltered, setIsStarredFiltered] = useState(false);

    const filteredRepos = repos
        .filter(repo => !isStarredFiltered || starred[repo.id])
        .filter(repo => language === null || (repo.language === language));

    return (
        <>
            <GithubTrendingHead />

            <Container maxWidth="md">
                <Headline>Github Trending</Headline>

                <FilterBar
                    isStarredFiltered={isStarredFiltered}
                    availableLanguages={repos.map(repo => repo.language).filter(repo => Boolean(repo))}
                    onChangeStarred={() => setIsStarredFiltered(c => !c)}
                    onChangeLanguage={setLanguage}
                />

                <GithubRepoList
                    filteredRepos={filteredRepos}
                    starred={starred}
                    onToggleStar={repoId => onToggleStar(repoId, repos.map(rp => rp.id))}
                />
            </Container>
        </>
    );
};

GithubTrendingPage.propTypes = {
    repos: PropTypes.arrayOf(githubRepoPropType).isRequired,
    starred: PropTypes.shape({}).isRequired,
    onToggleStar: PropTypes.func.isRequired,
};

export async function getStaticProps() {
    const dateOneWeekAgo = format(addWeeks(new Date(), -1), 'yyyy-MM-dd');

    const res = await fetch(`https://api.github.com/search/repositories?q=created:%3E${dateOneWeekAgo}&sort=stars&order=desc`);
    const repos = await res.json();

    return {
        props: {
            // This mapping saves space in the html file, since the api request is so big and the props get appended automatically by nextjs!
            // Although it is a little lengthy, it is hard to think of a better solution. Moving it into separat file feels like putting too
            // much space between the solution and the problem. Also using pick() from lodash/ramda feels like overkill.
            repos: repos.items?.map(({
                id,
                name,
                // eslint-disable-next-line camelcase
                html_url,
                description,
                watchers,
                language,
            }) => ({
                id,
                name,
                html_url,
                description,
                watchers,
                language,
            })),
        },
        // stale-while-revalidate, see https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration
        revalidate: 10,
    };
}

export default connect(
    state => ({ ...state.githubTrending }),
    dispatch => bindActionCreators({
        onToggleStar: toggleStar,
    }, dispatch),

)(GithubTrendingPage);
