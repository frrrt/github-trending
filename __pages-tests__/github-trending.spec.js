import React from 'react';

import {
    render,
    fireEvent,
    screen,
} from '@testing-library/react';
import { GithubTrendingPage } from '../pages/github-trending';

const baseProps = {
    repos: [
        {
            id: 300592907,
            name: 'GHunt',
            html_url: 'https://github.com/mxrch/GHunt',
            description: '🕵️‍♂️ Investigate Google Accounts with emails. ',
            watchers: 2950,
            language: 'Python',
        },
        {
            id: 300996055,
            name: 'vit-pytorch',
            html_url: 'https://github.com/lucidrains/vit-pytorch',
            description: 'Implementation of Vision Transformer, a simple way to achieve SOTA in vision classification with only a single transformer encoder, in Pytorch',
            watchers: 299,
            language: 'Python',
        },
        {
            id: 300450780,
            name: 'swift-atomics',
            html_url: 'https://github.com/apple/swift-atomics',
            description: null,
            watchers: 244,
            language: 'Swift',
        },
        {
            id: 301131109,
            name: 'mtproto',
            html_url: 'https://github.com/xelaj/mtproto',
            description: 'Full-native go implementation of Telegram API',
            watchers: 244,
            language: 'Go',
        },
    ],
    starred: {
        300450780: true,
        300996055: true,
    },
    onToggleStar: jest.fn(),
};

describe('GithubTrendingPage', () => {
    it('is rendered correctly', () => {
        expect(render(<GithubTrendingPage {...baseProps} />).container).toMatchSnapshot();
    });

    it('renders empty state correctly', () => {
        expect(render(<GithubTrendingPage {...baseProps} repos={[]} starred={{}} />).container).toMatchSnapshot();
    });

    it('calls onToggleStar correctly', () => {
        render(<GithubTrendingPage {...baseProps} />);

        const button = screen.getByText('2950');
        fireEvent.click(button);
        expect(baseProps.onToggleStar.mock.calls.length).toEqual(1);
        expect(baseProps.onToggleStar.mock.calls[0][0]).toEqual(300592907);

        fireEvent.click(button);
        expect(baseProps.onToggleStar.mock.calls.length).toEqual(2);
        expect(baseProps.onToggleStar.mock.calls[1][0]).toEqual(300592907);
    });

    it('toggles starred correctly', () => {
        render(<GithubTrendingPage {...baseProps} />);

        const button = screen.getByText('Starred');
        fireEvent.click(button);

        expect(screen.queryByText('2950')).toBeNull();
    });

    it('filter language', () => {
        render(<GithubTrendingPage {...baseProps} />);

        fireEvent.change(screen.getByRole('textbox'), { target: { value: 'Python' } });
        fireEvent.keyDown(screen.getByRole('textbox'), { keyCode: 13 });

        fireEvent.click(screen.getByRole('option'));

        expect(screen.queryByText('2950')).not.toBeNull();
        expect(screen.queryByText('244')).toBeNull();
    });
});
