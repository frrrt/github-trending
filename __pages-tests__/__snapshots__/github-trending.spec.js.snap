// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubTrendingPage is rendered correctly 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd"
  >
    <div
      class="MuiBox-root MuiBox-root-1"
    >
      <h1
        class="MuiTypography-root MuiTypography-h2"
      >
        Github Trending
      </h1>
    </div>
    <div
      class="MuiBox-root MuiBox-root-5 makeStyles-root-2"
    >
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root makeStyles-root-6 MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
        role="combobox"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink="false"
            for="language-selector"
            id="language-selector-label"
          >
            Language
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          >
            <input
              aria-autocomplete="list"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              id="language-selector"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="Clear"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-9"
              >
                <span>
                  Language
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-starredButton-3"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-starredIcon-4"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
            />
          </svg>
           Starred
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiBox-root MuiBox-root-12"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root makeStyles-row-11 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-4"
        >
          <div
            class="MuiPaper-root MuiCard-root makeStyles-root-13 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiCardContent-root"
            >
              <div
                class="MuiTypography-root makeStyles-repoTitle-14 MuiTypography-h6 MuiTypography-gutterBottom"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                  href="https://github.com/mxrch/GHunt"
                  rel="noopener noreferrer"
                  target="_blank"
                  title="GHunt"
                >
                  GHunt
                </a>
              </div>
              <div
                class="MuiTypography-root makeStyles-main-15 MuiTypography-body2 MuiTypography-gutterBottom"
              >
                <div
                  class="makeStyles-root-16"
                >
                  <div
                    class="makeStyles-circle-17"
                    style="background-color: rgb(53, 114, 165);"
                  />
                  Python
                </div>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    <span
                      class="MuiButton-startIcon MuiButton-iconSizeSmall"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19.65 9.04l-4.84-.42-1.89-4.45c-.34-.81-1.5-.81-1.84 0L9.19 8.63l-4.83.41c-.88.07-1.24 1.17-.57 1.75l3.67 3.18-1.1 4.72c-.2.86.73 1.54 1.49 1.08l4.15-2.5 4.15 2.51c.76.46 1.69-.22 1.49-1.08l-1.1-4.73 3.67-3.18c.67-.58.32-1.68-.56-1.75zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                        />
                      </svg>
                    </span>
                    2950
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
              >
                üïµÔ∏è‚Äç‚ôÇÔ∏è Investigate Google Accounts with emails. 
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root makeStyles-row-11 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-4"
        >
          <div
            class="MuiPaper-root MuiCard-root makeStyles-root-13 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiCardContent-root"
            >
              <div
                class="MuiTypography-root makeStyles-repoTitle-14 MuiTypography-h6 MuiTypography-gutterBottom"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                  href="https://github.com/lucidrains/vit-pytorch"
                  rel="noopener noreferrer"
                  target="_blank"
                  title="vit-pytorch"
                >
                  vit-pytorch
                </a>
              </div>
              <div
                class="MuiTypography-root makeStyles-main-15 MuiTypography-body2 MuiTypography-gutterBottom"
              >
                <div
                  class="makeStyles-root-16"
                >
                  <div
                    class="makeStyles-circle-17"
                    style="background-color: rgb(53, 114, 165);"
                  />
                  Python
                </div>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    <span
                      class="MuiButton-startIcon MuiButton-iconSizeSmall"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27l4.15 2.51c.76.46 1.69-.22 1.49-1.08l-1.1-4.72 3.67-3.18c.67-.58.31-1.68-.57-1.75l-4.83-.41-1.89-4.46c-.34-.81-1.5-.81-1.84 0L9.19 8.63l-4.83.41c-.88.07-1.24 1.17-.57 1.75l3.67 3.18-1.1 4.72c-.2.86.73 1.54 1.49 1.08l4.15-2.5z"
                        />
                      </svg>
                    </span>
                    300
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
              >
                Implementation of Vision Transformer, a simple way to achieve SOTA in vision classification with only a single transformer encoder, in Pytorch
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root makeStyles-row-11 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-4"
        >
          <div
            class="MuiPaper-root MuiCard-root makeStyles-root-13 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiCardContent-root"
            >
              <div
                class="MuiTypography-root makeStyles-repoTitle-14 MuiTypography-h6 MuiTypography-gutterBottom"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                  href="https://github.com/apple/swift-atomics"
                  rel="noopener noreferrer"
                  target="_blank"
                  title="swift-atomics"
                >
                  swift-atomics
                </a>
              </div>
              <div
                class="MuiTypography-root makeStyles-main-15 MuiTypography-body2 MuiTypography-gutterBottom"
              >
                <div
                  class="makeStyles-root-16"
                >
                  <div
                    class="makeStyles-circle-17"
                    style="background-color: rgb(255, 172, 69);"
                  />
                  Swift
                </div>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    <span
                      class="MuiButton-startIcon MuiButton-iconSizeSmall"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27l4.15 2.51c.76.46 1.69-.22 1.49-1.08l-1.1-4.72 3.67-3.18c.67-.58.31-1.68-.57-1.75l-4.83-.41-1.89-4.46c-.34-.81-1.5-.81-1.84 0L9.19 8.63l-4.83.41c-.88.07-1.24 1.17-.57 1.75l3.67 3.18-1.1 4.72c-.2.86.73 1.54 1.49 1.08l4.15-2.5z"
                        />
                      </svg>
                    </span>
                    245
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root makeStyles-row-11 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-4"
        >
          <div
            class="MuiPaper-root MuiCard-root makeStyles-root-13 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiCardContent-root"
            >
              <div
                class="MuiTypography-root makeStyles-repoTitle-14 MuiTypography-h6 MuiTypography-gutterBottom"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                  href="https://github.com/xelaj/mtproto"
                  rel="noopener noreferrer"
                  target="_blank"
                  title="mtproto"
                >
                  mtproto
                </a>
              </div>
              <div
                class="MuiTypography-root makeStyles-main-15 MuiTypography-body2 MuiTypography-gutterBottom"
              >
                <div
                  class="makeStyles-root-16"
                >
                  <div
                    class="makeStyles-circle-17"
                    style="background-color: rgb(0, 173, 216);"
                  />
                  Go
                </div>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    <span
                      class="MuiButton-startIcon MuiButton-iconSizeSmall"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19.65 9.04l-4.84-.42-1.89-4.45c-.34-.81-1.5-.81-1.84 0L9.19 8.63l-4.83.41c-.88.07-1.24 1.17-.57 1.75l3.67 3.18-1.1 4.72c-.2.86.73 1.54 1.49 1.08l4.15-2.5 4.15 2.51c.76.46 1.69-.22 1.49-1.08l-1.1-4.73 3.67-3.18c.67-.58.32-1.68-.56-1.75zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                        />
                      </svg>
                    </span>
                    244
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
              >
                Full-native go implementation of Telegram API
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GithubTrendingPage renders empty state correctly 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd"
  >
    <div
      class="MuiBox-root MuiBox-root-18"
    >
      <h1
        class="MuiTypography-root MuiTypography-h2"
      >
        Github Trending
      </h1>
    </div>
    <div
      class="MuiBox-root MuiBox-root-22 makeStyles-root-19"
    >
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root makeStyles-root-23 MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
        role="combobox"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink="false"
            for="language-selector"
            id="language-selector-label"
          >
            Language
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          >
            <input
              aria-autocomplete="list"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              id="language-selector"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="Clear"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-24 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-26"
              >
                <span>
                  Language
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-starredButton-20"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-starredIcon-21"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
            />
          </svg>
           Starred
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiBox-root MuiBox-root-29"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      />
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        No repos found!
      </p>
    </div>
  </div>
</div>
`;
